torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,352 @@
+            output_device = device_ids[0]
+
+        self.dim = dim
+        self.module = module
+        self.process_group = process_group
+        self.device_ids = device_ids
+        self.output_device = output_device
+        self.broadcast_buffers = broadcast_buffers
+
+        MB = 1024 * 1024
+
+        # used for intra-node param sync and inter-node sync as well
+        self.broadcast_bucket_size = 25 * MB
+
+        # Sync params and buffers
+        module_states = list(self.module.state_dict().values())
+        if len(module_states) > 0:
+            self._dist_broadcast_coalesced(module_states,
+                                           self.broadcast_bucket_size)
+
+        if len(device_ids) > 1:
+            # TODO: we don't need to replicate params in here. they're always going to
+            # be broadcasted using larger blocks in broadcast_coalesced, so it might be
+            # better to not pollute the caches with these small blocks
+            self._module_copies = replicate(self.module, self.device_ids, detach=True)
+            self._module_copies[0] = self.module
+
+            for module_copy in self._module_copies[1:]:
+                for param, copy_param in zip(self.module.parameters(), module_copy.parameters()):
+                    copy_param.requires_grad = param.requires_grad
+
+        else:
+            self._module_copies = [self.module]
+
+        bucket_bytes_cap = 25 * MB
+
+        # This is a triply-nested list where the "dimensions" are: devices, buckets, bucket_elems
+        param_buckets = []
+        # Split the parameters into buckets and by types as well
+        for dev_idx, module in enumerate(self._module_copies):
+            param_buckets.append(list(_take_tensors(module.parameters(), bucket_bytes_cap)))
+
+        self.bucket_sizes = []
+        self.bucket_map = {}
+
+        # We transpose param_buckets, so the loop is over buckets.
+        # param_buckets_tuple is a doubly-nested list with "dims": devices, bucket_elems
+        for bucket_idx, param_buckets_tuple in enumerate(zip(*param_buckets)):
+            self.bucket_sizes.append(0)
+            # Now, we transpose again, so we iterate over bucket_elems, but getting tuples
+            # of params from each device.
+            for idx, param_tuple in enumerate(zip(*param_buckets_tuple)):
+                if not param_tuple[0].requires_grad:
+                    continue
+                for p in param_tuple:
+                    self.bucket_map[p] = bucket_idx
+                self.bucket_sizes[bucket_idx] += 1
+
+        self.buckets = [[[] for _ in range(len(self.device_ids))] for _ in range(len(self.bucket_sizes))]
+        # coalesced bucket for only device 0
+        self.buckets_coalesced = [[] for _ in range(len(self.bucket_sizes))]
+
+        # We will always reduce the bucket following the reverse order
+        self.bucket_order = list(range(len(self.bucket_sizes)))
+        self.bucket_order.reverse()
```

Comment:
`bucket_order` is simply `n`...`0`. Do we really need to materialize it?

Reply:
we wouldn't have many buckets anyway, and it's simple to understand. So I guess it's fine.

Reply:
Removed the reverse